app/filtering.js

```diff
@@ -699,6 +759,72 @@ const initPartition = (partition) => {
 }
 module.exports.initPartition = initPartition
 
+function setupTor () {
+  let torInitialized = null
+  // If Tor has not successfully initialized or thrown an error within 20s,
+  // assume it's broken.
+  setTimeout(() => {
+    if (torInitialized === null) {
+      appActions.onTorInitError(`Tor could not start.`)
+    }
+  }, 20000)
+  // Set up the tor daemon watcher.  (NOTE: We don't actually start
+  // the tor daemon here; that happens in C++ code.  But we do talk to
+  // its control socket.)
+  const torDaemon = new tor.TorDaemon()
+  torDaemon.setup((err) => {
+    if (err) {
+      appActions.onTorInitError(`Tor failed to make directories: ${err}`)
+      torInitialized = false
+      return
+    }
+    torDaemon.on('exit', () => {
+      appActions.onTorInitError('The Tor process has stopped.')
+      torInitialized = false
+    })
+    torDaemon.on('launch', (socksAddr) => {
+      console.log(`tor: daemon listens on ${socksAddr}`)
+      const bootstrapped = (err, progress) => {
+        if (err) {
+          appActions.onTorInitError(`Tor bootstrap error: ${err}`)
+          torInitialized = false
+          return
+        }
+        appActions.onTorInitPercentage(progress)
+      }
+      const circuitEstablished = (err, ok) => {
+        if (ok) {
+          console.log('Tor ready!')
+          appActions.onTorInitSuccess()
+          torInitialized = true
+        } else {
+          if (err) {
+            appActions.onTorInitError(`Tor not ready: ${err}`)
+            torInitialized = false
+          } else {
+            // Simply log the error but don't show error UI since Tor might
+            // finish opening a circuit.
+            console.log('tor still not ready')
+          }
+        }
+      }
+      torDaemon.onBootstrap(bootstrapped, (err) => {
+        if (err) {
+          appActions.onTorInitError(`Tor error bootstrapping: ${err}`)
+          torInitialized = false
+        }
+        torDaemon.onCircuitEstablished(circuitEstablished, (err) => {
+          if (err) {
+            appActions.onTorInitError(`Tor error opening a circuit: ${err}`)
```

Comment:
Just FYI: An error here means not that there was a problem opening a circuit, but that there was a problem in _subscribing to the circuit establishment notifications_.  (It's rather unlikely that this would happen.)